core/domain/exp_services.py

```diff
@@ -568,52 +569,56 @@ def _save_exploration(committer_id, exploration, commit_message, change_list):
-            'state_names_with_changed_answer_groups']
-        state_names_with_unchanged_answer_groups = trainable_states_dict[
-            'state_names_with_unchanged_answer_groups']
-        state_names_to_train_classifier = state_names_with_changed_answer_groups
-        if state_names_with_unchanged_answer_groups:
-            state_names_without_classifier = (
-                classifier_services.handle_non_retrainable_states(
-                    exploration, state_names_with_unchanged_answer_groups,
-                    exp_versions_diff))
-            state_names_to_train_classifier.extend(
-                state_names_without_classifier)
-        if state_names_to_train_classifier:
-            classifier_services.handle_trainable_states(
-                exploration, state_names_to_train_classifier)
-
-    # Trigger exploration issues model updation.
-    stats_services.update_exp_issues_for_new_exp_version(
-        exploration, exp_versions_diff=exp_versions_diff,
-        revert_to_version=None)
+    def update_storage_models():
+        """Commits all the newly created models in an atomic transaction. If any
+        operation fails, every other operation will fail as well.
+        """
+        exploration_model.commit(committer_id, commit_message, change_list_dict)
+        exploration.version += 1
+
+        if feconf.ENABLE_ML_CLASSIFIERS:
+            trainable_states_dict = exploration.get_trainable_states_dict(
+                old_states, exp_versions_diff)
+            state_names_with_changed_answer_groups = trainable_states_dict[
+                'state_names_with_changed_answer_groups']
+            state_names_with_unchanged_answer_groups = trainable_states_dict[
+                'state_names_with_unchanged_answer_groups']
+            state_names_to_train_classifier = (
+                state_names_with_changed_answer_groups)
+            if state_names_with_unchanged_answer_groups:
+                state_names_without_classifier = (
+                    classifier_services.handle_non_retrainable_states(
+                        exploration, state_names_with_unchanged_answer_groups,
+                        exp_versions_diff))
+                state_names_to_train_classifier.extend(
+                    state_names_without_classifier)
+            if state_names_to_train_classifier:
+                classifier_services.handle_trainable_states(
+                    exploration, state_names_to_train_classifier)
+
+        stats_services.create_stats_model(
+            stats_services.get_stats_for_new_exp_version(
+                exploration.id, exploration.version, exploration.states,
+                exp_versions_diff=exp_versions_diff, revert_to_version=None))
+
+        stats_services.update_exp_issues_for_new_exp_version(
+            exploration, exp_versions_diff=exp_versions_diff,
+            revert_to_version=None)
+
+        memcache_services.delete(
```

Comment:
Does memcache stuff run in a transaction as well? (I.e. does it get rolled back if the transaction fails?)

If not -- should only do it outside the transaction, after it is run?

Reply:
I decided to leave it out of the transaction since it didn't seem critical to logic, so not sure how it interacts with transactional stuff. Should we add it in?